ingester/ingester.go

```diff
@@ -487,42 +489,70 @@ func (i *Ingester) flushAllUsers(immediate bool) {
 	i.userStateLock.Unlock()
 
 	for id, state := range userState {
-		i.flushUser(id, state, immediate)
-	}
-}
-
-func (i *Ingester) flushUser(userID string, userState *userState, immediate bool) {
-	for pair := range userState.fpToSeries.iter() {
-		i.flushSeries(userState, pair.fp, pair.series, immediate)
+		for pair := range state.fpToSeries.iter() {
+			state.fpLocker.Lock(pair.fp)
+			i.sweepSeries(id, pair.fp, pair.series, immediate)
+			state.fpLocker.Unlock(pair.fp)
+		}
 	}
 
-	// TODO: this is probably slow, and could be done in a better way.
 	i.userStateLock.Lock()
-	if userState.fpToSeries.length() == 0 {
-		delete(i.userState, userID)
+	for id, state := range userState {
+		if state.fpToSeries.length() == 0 {
+			delete(i.userState, id)
+		}
```

Comment:
> It would. But no such code exists.

Cool. Just wanted to confirm my understanding.

> file a separate issue and work on fixing that next

SGTM